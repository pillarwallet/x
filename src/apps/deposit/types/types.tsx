export type BalanceInfo = {
  type?: 'BalanceInfo';
  name: string;
  chain: string;
  address: string;
  decimals: number;
  symbol: string;
  logoURI: string;
  balance: string;
};

export type AddedAssets = {
  type?: 'AddedAsset';
  assetType: 'token' | 'nft';
  tokenId?: string;
  chain: string;
  tokenAddress: string;
  balance: number;
};

export type Network =
  | 'ethereum'
  | 'polygon'
  | 'gnosis'
  | 'base'
  | 'bnb smart chain'
  | 'optimism'
  | 'arbitrum';

export type TokenList = {
  address: string;
  chainId: number;
  decimals: string;
  name: string;
  symbol: string;
  logoURI: string;
};

export type BalancesByChain = {
  [chain: string]: BalanceInfo[];
};
