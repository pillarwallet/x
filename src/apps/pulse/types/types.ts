export type TransactionStatusState =
  | 'Starting Transaction'
  | 'Transaction Pending'
  | 'Transaction Complete'
  | 'Transaction Failed';

export type TransactionStep =
  | 'Submitted'
  | 'Pending'
  | 'ResourceLock'
  | 'Completed';

export type StepStatus = 'completed' | 'pending' | 'failed' | 'inactive';

export interface TokenInfo {
  symbol: string;
  name: string;
  logo: string;
  address?: string;
}

export interface SellOffer {
  tokenAmountToReceive: number;
  minimumReceive: number;
}

export interface PayingToken {
  totalUsd: number;
  name: string;
  symbol: string;
  logo: string;
  actualBal: string;
  totalRaw: string;
  chainId: number;
  address: string;
}

export interface TokenDetails {
  symbol: string;
  name: string;
  address: string;
  chainId: number;
  amount: string;
  logo: string;
  type: 'BUY_TOKEN' | 'SELL_TOKEN';
}

export interface BuyModeDetails {
  usdAmount: string;
  payingTokens: PayingToken[];
  totalPayingUsd: number;
}

export interface TransactionStatusConfig {
  icon: string;
  containerClasses: string;
  iconClasses: string;
  color: string;
}

export interface ButtonConfig {
  bgColor: string;
  textColor: string;
  borderColor: string;
  label: string;
}

export interface ProgressStepConfig {
  label: string;
  order: number;
}

export interface TransactionStatusProps {
  closeTransactionStatus: () => void;
  userOpHash: string;
  chainId: number;
  gasFee?: string;
  // Transaction data from PreviewSell/PreviewBuy
  isBuy?: boolean;
  sellToken?: TokenInfo | null;
  buyToken?: TokenInfo | null;
  tokenAmount?: string;
  sellOffer?: SellOffer | null;
  payingTokens?: PayingToken[];
  usdAmount?: string;
  // Externalized polling state
  currentStatus: TransactionStatusState;
  errorDetails: string;
  submittedAt?: Date;
  pendingCompletedAt?: Date;
  blockchainTxHash?: string;
  resourceLockTxHash?: string;
  completedTxHash?: string;
  completedChainId?: number;
  resourceLockChainId?: number;
  resourceLockCompletedAt?: Date;
  isResourceLockFailed?: boolean;
}

export interface TransactionDetailsProps {
  onDone: () => void;
  userOpHash: string;
  chainId: number;
  status: TransactionStatusState;
  // Transaction data from PreviewSell/PreviewBuy
  isBuy?: boolean;
  sellToken?: TokenInfo | null;
  buyToken?: TokenInfo | null;
  tokenAmount?: string;
  sellOffer?: SellOffer | null;
  payingTokens?: PayingToken[];
  usdAmount?: string;
  submittedAt?: Date;
  pendingCompletedAt?: Date;
  resourceLockCompletedAt?: Date;
  txHash?: string;
  gasFee?: string;
  errorDetails?: string;
  resourceLockTxHash?: string;
  completedTxHash?: string;
  resourceLockChainId?: number;
  completedChainId?: number;
  isResourceLockFailed?: boolean;
}

export interface TransactionInfoProps {
  status: TransactionStatusState;
  userOpHash: string;
  txHash?: string;
  chainId: number;
  gasFee?: string;
  completedAt?: Date;
  // Buy-specific fields
  isBuy?: boolean;
  resourceLockTxHash?: string;
  resourceLockChainId?: number;
  completedTxHash?: string;
  completedChainId?: number;
}

export interface ProgressStepProps {
  step: TransactionStep;
  status: StepStatus;
  label: string;
  isLast?: boolean;
  showLine?: boolean;
  lineStatus?: StepStatus;
  timestamp?: string | React.ReactNode;
}

export interface TransactionErrorBoxProps {
  technicalDetails?: string;
}

export interface UseClickOutsideOptions {
  ref: React.RefObject<HTMLElement>;
  callback: () => void;
  condition: boolean;
}

export interface UseKeyboardNavigationOptions {
  onEscape: () => void;
  onEnter?: () => void;
  enabled?: boolean;
}

export interface TechnicalDetails {
  transactionType: 'BUY' | 'SELL';
  transactionHash: string;
  hashType: 'bidHash' | 'userOpHash';
  chainId: number;
  status: TransactionStatusState;
  timestamp: string;
  accountAddress: string;
  token: TokenDetails | null;
  sellOffer: SellOffer | null;
  buyMode: BuyModeDetails | null;
  transactionHashes: {
    [key: string]: string;
  };
  chains: {
    mainChainId: number;
    resourceLockChainId: string | number;
    completedChainId: string | number;
  };
  timestamps: {
    [key: string]: string;
  };
  error: {
    details: string;
    isResourceLockFailed: boolean;
    failureStep: string;
  };
  gas: {
    fee: string;
  };
  stepStatus: {
    [key: string]: StepStatus;
  };
}
