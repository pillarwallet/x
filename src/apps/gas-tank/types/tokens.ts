export type SelectedToken = {
  name: string;
  symbol: string;
  usdValue: string;
  logo: string;
  dailyPriceChange: number;
  chainId: number;
  decimals: number;
  address: string;
};

export type PayingToken = {
  name: string;
  symbol: string;
  logo: string;
  actualBal: string;
  totalUsd: number;
  totalRaw: string;
  chainId: number;
  address: string;
};

export enum SearchType {
  MyHoldings = 'My Holdings',
}

export enum SortType {
  Up = 'Up',
  Down = 'Down',
}
