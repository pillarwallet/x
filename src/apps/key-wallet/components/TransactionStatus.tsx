import { TransactionStatus as TxStatus } from '../types';
import { getBlockExplorerUrl, shortenAddress } from '../utils/blockchain';

interface TransactionStatusProps {
  transactions: TxStatus[];
  onClear: (hash: string) => void;
}

const TransactionStatus = ({
  transactions,
  onClear,
}: TransactionStatusProps) => {
  if (transactions.length === 0) return null;

  return (
    <div className="w-full mb-6">
      <h2 className="text-lg font-semibold text-white mb-4">
        Recent Transactions
      </h2>
      <div className="space-y-3">
        {transactions.map((tx) => {
          const explorerUrl = getBlockExplorerUrl(tx.chainId, tx.hash);
          const statusColor =
            tx.status === 'success'
              ? 'text-green-400'
              : tx.status === 'failed'
                ? 'text-red-400'
                : 'text-yellow-400';
          const statusIcon =
            tx.status === 'success' ? '✓' : tx.status === 'failed' ? '✗' : '⏳';

          return (
            <div
              key={tx.hash}
              className="bg-white/5 rounded-xl p-4 border border-white/10"
            >
              <div className="flex items-start justify-between gap-4">
                <div className="flex-1 min-w-0">
                  <div className="flex items-center gap-2 mb-2">
                    <span className={`text-lg ${statusColor}`}>
                      {statusIcon}
                    </span>
                    <span className={`text-sm font-medium ${statusColor}`}>
                      {tx.status === 'pending'
                        ? 'Transaction Pending'
                        : tx.status === 'success'
                          ? 'Transaction Confirmed'
                          : 'Transaction Failed'}
                    </span>
                  </div>
                  <p className="text-xs text-white/60 font-mono break-all">
                    {shortenAddress(tx.hash, 8)}
                  </p>
                  <p className="text-xs text-white/40 mt-1">
                    {new Date(tx.timestamp).toLocaleString()}
                  </p>
                </div>
                <div className="flex gap-2 flex-shrink-0">
                  {explorerUrl && (
                    <a
                      href={explorerUrl}
                      target="_blank"
                      rel="noopener noreferrer"
                      className="px-3 py-1 bg-purple_medium hover:bg-purple_light rounded-lg text-xs font-medium transition-colors"
                    >
                      View
                    </a>
                  )}
                  <button
                    onClick={() => onClear(tx.hash)}
                    className="px-3 py-1 bg-white/5 hover:bg-white/10 rounded-lg text-xs font-medium transition-colors"
                    type="button"
                  >
                    Clear
                  </button>
                </div>
              </div>
            </div>
          );
        })}
      </div>
    </div>
  );
};

export default TransactionStatus;

