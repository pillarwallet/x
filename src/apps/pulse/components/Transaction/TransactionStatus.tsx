import { animated, useSpring } from '@react-spring/web';
import { useEffect, useRef, useState } from 'react';

// components
import TransactionDetails from './TransactionDetails';
import TransactionStatusContainer from './TransactionStatusContainer';

// hooks
import { useClickOutside } from '../../hooks/useClickOutside';
import { useKeyboardNavigation } from '../../hooks/useKeyboardNavigation';

// types
import { TransactionStatusProps } from '../../types/types';

const TransactionStatus = (props: TransactionStatusProps) => {
  const {
    closeTransactionStatus,
    userOpHash,
    chainId,
    gasFee,
    isBuy,
    sellToken,
    buyToken,
    tokenAmount,
    sellOffer,
    payingTokens,
    usdAmount,
    // Externalized polling state
    currentStatus,
    errorDetails,
    submittedAt,
    pendingCompletedAt,
    blockchainTxHash,
    resourceLockTxHash,
    completedTxHash,
    completedChainId,
    resourceLockChainId,
    resourceLockCompletedAt,
    isResourceLockFailed,
  } = props;

  const [showDetails, setShowDetails] = useState(false);
  const transactionStatusRef = useRef<HTMLDivElement>(null);

  // Bouncing animation spring
  const [springs, api] = useSpring(() => ({
    from: {
      scale: 1,
      opacity: 1,
      transformOrigin: 'center center',
    },
    to: {
      scale: 1,
      opacity: 1,
      transformOrigin: 'center center',
    },
  }));

  // Trigger bouncing animation when showDetails changes
  useEffect(() => {
    // Quick bounce effect when size changes
    api.start({
      to: {
        scale: 1.04,
        opacity: 1,
        transformOrigin: 'center center',
      },
      config: {
        duration: 100,
        tension: 400,
        friction: 10,
      },
      onRest: () => {
        // Bounce back to normal size
        api.start({
          to: {
            scale: 1,
            opacity: 1,
            transformOrigin: 'center center',
          },
          config: {
            duration: 120,
            tension: 300,
            friction: 25,
          },
        });
      },
    });
  }, [showDetails, api]);

  // Auto-show details when transaction completes or fails (with 1 second delay)
  useEffect(() => {
    if (
      (currentStatus === 'Transaction Complete' ||
        currentStatus === 'Transaction Failed') &&
      !showDetails
    ) {
      const timer = setTimeout(() => {
        setShowDetails(true);
      }, 1000);

      return () => clearTimeout(timer);
    }
    return undefined;
  }, [currentStatus, showDetails]);

  useClickOutside({
    ref: transactionStatusRef,
    callback: closeTransactionStatus,
    condition:
      currentStatus === 'Transaction Complete' ||
      currentStatus === 'Transaction Failed',
  });

  useKeyboardNavigation({
    onEscape: () => {
      if (
        currentStatus === 'Transaction Complete' ||
        currentStatus === 'Transaction Failed'
      ) {
        closeTransactionStatus();
      } else if (currentStatus === 'Transaction Pending' && showDetails) {
        setShowDetails(false);
      }
    },
  });

  return (
    <animated.div
      ref={transactionStatusRef}
      className={`flex flex-col w-full max-w-[446px] ${!showDetails ? 'h-[calc(100vh-94px)] max-h-[600px]' : 'h-min'} bg-[#1E1D24] border border-white/5 rounded-[10px] p-6`}
      style={springs}
      data-testid="pulse-transaction-status"
    >
      {showDetails ? (
        <TransactionDetails
          onDone={
            currentStatus === 'Transaction Pending'
              ? () => setShowDetails(false)
              : closeTransactionStatus
          }
          userOpHash={userOpHash}
          chainId={chainId}
          status={currentStatus}
          isBuy={isBuy}
          sellToken={sellToken}
          buyToken={buyToken}
          tokenAmount={tokenAmount}
          sellOffer={sellOffer}
          payingTokens={payingTokens}
          usdAmount={usdAmount}
          submittedAt={submittedAt}
          pendingCompletedAt={pendingCompletedAt}
          txHash={isBuy ? undefined : blockchainTxHash}
          gasFee={gasFee}
          errorDetails={errorDetails}
          resourceLockTxHash={resourceLockTxHash}
          completedTxHash={isBuy ? completedTxHash : undefined}
          resourceLockChainId={resourceLockChainId}
          completedChainId={completedChainId}
          resourceLockCompletedAt={resourceLockCompletedAt}
          isResourceLockFailed={isResourceLockFailed}
        />
      ) : (
        <TransactionStatusContainer
          status={currentStatus}
          onViewDetails={() => setShowDetails(true)}
        />
      )}
    </animated.div>
  );
};

export default TransactionStatus;
