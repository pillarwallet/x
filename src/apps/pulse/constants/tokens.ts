export const STABLE_CURRENCIES = [
  { chainId: 10, address: '0x0b2C639c533813f4Aa9D7837CAf62653d097Ff85' }, // USDC on Optimism
  { chainId: 137, address: '0x3c499c542cEF5E3811e1192ce70d8cC03d5c3359' }, // USDC on Polygon
  { chainId: 8453, address: '0x833589fCD6eDb6E08f4c7C32D4f71b54bdA02913' }, // USDC on Base
  { chainId: 42161, address: '0xaf88d065e77c8cC2239327C5EDb3A432268e5831' }, // USDC on Arbitrum
  { chainId: 56, address: '0x8AC76a51cc950d9822D68b83fE1Ad97B32Cd580d' }, // USDC on BNB Smart Chain
  { chainId: 100, address: '0x2a22f9c3b484c3629090FeED35F17Ff8F88f76F0' }, // USDC on Gnosis
];
