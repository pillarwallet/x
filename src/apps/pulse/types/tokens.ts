type ParsedToken = {
  name: string;
  symbol: string;
  markrtCap: number;
  liquidity: number;
  volume: number;
  logo: string;
  price: number;
  priceChange: number;
  blockchain: string;
}

type SelectedToken = {
  name: string;
  symbol: string;
  usdValue: string;
  logo: string;
  dailyPriceChange: number;
  chainId: number;
  decimals: number;
  address: string;
}