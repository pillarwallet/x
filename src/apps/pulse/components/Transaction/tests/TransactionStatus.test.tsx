/* eslint-disable react/jsx-props-no-spreading */
/* eslint-disable @typescript-eslint/no-explicit-any */
import { fireEvent, render, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';
import { vi } from 'vitest';

// types
import { SelectedToken } from '../../../types/tokens';

// components
import TransactionStatus from '../TransactionStatus';

// Mock dependencies
vi.mock('../../assets/confirmed-icon.svg', () => ({
  default: 'confirmed-icon.svg',
}));

vi.mock('../../assets/failed-icon.svg', () => ({
  default: 'failed-icon.svg',
}));

vi.mock('../../assets/pending.svg', () => ({
  default: 'pending.svg',
}));

vi.mock('../TransactionDetails', () => ({
  default: ({ onDone }: { onDone: () => void }) => (
    <div data-testid="transaction-details">
      <button type="button" onClick={onDone} data-testid="done-button">
        Done
      </button>
    </div>
  ),
}));

const mockSellToken: SelectedToken = {
  name: 'Test Token',
  symbol: 'TEST',
  logo: 'test-logo.png',
  usdValue: '100.00',
  dailyPriceChange: 0.05,
  chainId: 1,
  decimals: 18,
  address: '0x1234567890123456789012345678901234567890',
};

const mockBuyToken: SelectedToken = {
  name: 'Buy Token',
  symbol: 'BUY',
  logo: 'buy-logo.png',
  usdValue: '50.00',
  dailyPriceChange: 0.02,
  chainId: 1,
  decimals: 18,
  address: '0x9876543210987654321098765432109876543210',
};

const mockPayingTokens = [
  {
    totalUsd: 100.0,
    name: 'USD Coin',
    symbol: 'USDC',
    logo: 'usdc-logo.png',
    actualBal: '100.00',
    totalRaw: '100000000',
    chainId: 1,
    address: '0xA0b86a33E6441b8C4C8C0C4C8C0C4C8C0C4C8C0C4',
  },
];

const mockSellOffer = {
  tokenAmountToReceive: 50.0,
  minimumReceive: 45.0,
};

const baseProps = {
  closeTransactionStatus: vi.fn(),
  userOpHash:
    '0x1234567890abcdef1234567890abcdef1234567890abcdef1234567890abcdef',
  chainId: 1,
  gasFee: '0.001',
  isBuy: false,
  sellToken: mockSellToken,
  buyToken: null,
  tokenAmount: '100',
  sellOffer: mockSellOffer,
  payingTokens: undefined,
  usdAmount: undefined,
  currentStatus: 'Starting Transaction' as const,
  errorDetails: '',
  submittedAt: new Date('2023-01-01T00:00:00Z'),
  pendingCompletedAt: undefined,
  blockchainTxHash:
    '0xabcdef1234567890abcdef1234567890abcdef1234567890abcdef1234567890',
  resourceLockTxHash: undefined,
  completedTxHash: undefined,
  completedChainId: undefined,
  resourceLockChainId: undefined,
  resourceLockCompletedAt: undefined,
  isResourceLockFailed: false,
};

describe('<TransactionStatus />', () => {
  beforeEach(() => {
    vi.clearAllMocks();
  });

  it('renders correctly and matches snapshot', () => {
    const tree = renderer.create(<TransactionStatus {...baseProps} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  describe('renders different transaction statuses', () => {
    it('displays starting transaction status correctly', () => {
      render(
        <TransactionStatus
          {...baseProps}
          currentStatus="Starting Transaction"
        />
      );

      expect(screen.getByText('Starting Transaction')).toBeInTheDocument();
      expect(screen.getByText('Just a moment...')).toBeInTheDocument();
      expect(screen.getByAltText('Starting Transaction')).toBeInTheDocument();
    });

    it('displays transaction pending status correctly', () => {
      render(
        <TransactionStatus {...baseProps} currentStatus="Transaction Pending" />
      );

      expect(screen.getByText('Transaction Pending')).toBeInTheDocument();
      expect(screen.getByText('View Status')).toBeInTheDocument();
      expect(screen.getByAltText('Transaction Pending')).toBeInTheDocument();
    });

    it('displays transaction complete status correctly', () => {
      render(
        <TransactionStatus
          {...baseProps}
          currentStatus="Transaction Complete"
        />
      );

      expect(screen.getByText('Transaction Complete')).toBeInTheDocument();
      expect(screen.getByText('Success')).toBeInTheDocument();
      expect(screen.getByAltText('Transaction Complete')).toBeInTheDocument();
    });

    it('displays transaction failed status correctly', () => {
      render(
        <TransactionStatus {...baseProps} currentStatus="Transaction Failed" />
      );

      expect(screen.getByText('Transaction Failed')).toBeInTheDocument();
      expect(screen.getByText('View Status')).toBeInTheDocument();
      expect(screen.getByAltText('Transaction Failed')).toBeInTheDocument();
    });
  });

  describe('handles user interactions', () => {
    it('opens details when view status button is clicked', () => {
      render(
        <TransactionStatus {...baseProps} currentStatus="Transaction Pending" />
      );

      const viewStatusButton = screen.getByText('View Status');
      fireEvent.click(viewStatusButton);

      expect(screen.getByTestId('transaction-details')).toBeInTheDocument();
    });

    it('opens details when success button is clicked', () => {
      render(
        <TransactionStatus
          {...baseProps}
          currentStatus="Transaction Complete"
        />
      );

      const successButton = screen.getByText('Success');
      fireEvent.click(successButton);

      expect(screen.getByTestId('transaction-details')).toBeInTheDocument();
    });

    it('opens details when failed view status button is clicked', () => {
      render(
        <TransactionStatus {...baseProps} currentStatus="Transaction Failed" />
      );

      const viewStatusButton = screen.getByText('View Status');
      fireEvent.click(viewStatusButton);

      expect(screen.getByTestId('transaction-details')).toBeInTheDocument();
    });

    it('returns to main view when done button is clicked in details for pending status', () => {
      render(
        <TransactionStatus {...baseProps} currentStatus="Transaction Pending" />
      );

      // Open details first
      const viewStatusButton = screen.getByText('View Status');
      fireEvent.click(viewStatusButton);

      // Click done button - should return to main view, not close
      const doneButton = screen.getByTestId('done-button');
      fireEvent.click(doneButton);

      expect(baseProps.closeTransactionStatus).not.toHaveBeenCalled();
      // Should be back to main view (no details shown)
      expect(screen.getByText('Transaction Pending')).toBeInTheDocument();
    });

    it('closes transaction status when done button is clicked in details for completed status', () => {
      render(
        <TransactionStatus
          {...baseProps}
          currentStatus="Transaction Complete"
        />
      );

      // Open details first
      const successButton = screen.getByText('Success');
      fireEvent.click(successButton);

      // Click done button - should close for completed status
      const doneButton = screen.getByTestId('done-button');
      fireEvent.click(doneButton);

      expect(baseProps.closeTransactionStatus).toHaveBeenCalled();
    });

    it('closes transaction status when ESC key is pressed for completed status', () => {
      render(
        <TransactionStatus
          {...baseProps}
          currentStatus="Transaction Complete"
        />
      );

      fireEvent.keyDown(document, { key: 'Escape' });

      expect(baseProps.closeTransactionStatus).toHaveBeenCalled();
    });

    it('does not close transaction status when ESC key is pressed for starting status', () => {
      render(
        <TransactionStatus
          {...baseProps}
          currentStatus="Starting Transaction"
        />
      );

      fireEvent.keyDown(document, { key: 'Escape' });

      expect(baseProps.closeTransactionStatus).not.toHaveBeenCalled();
    });

    it('returns to main view when ESC key is pressed in details for pending status', () => {
      render(
        <TransactionStatus {...baseProps} currentStatus="Transaction Pending" />
      );

      // Open details first
      const viewStatusButton = screen.getByText('View Status');
      fireEvent.click(viewStatusButton);

      // Press ESC - should return to main view
      fireEvent.keyDown(document, { key: 'Escape' });

      expect(baseProps.closeTransactionStatus).not.toHaveBeenCalled();
      // Should be back to main view
      expect(screen.getByText('Transaction Pending')).toBeInTheDocument();
    });

    it('closes transaction status when clicking outside the modal for completed status', () => {
      render(
        <TransactionStatus
          {...baseProps}
          currentStatus="Transaction Complete"
        />
      );

      // Click outside the modal
      fireEvent.mouseDown(document.body);

      expect(baseProps.closeTransactionStatus).toHaveBeenCalled();
    });

    it('does not close transaction status when clicking outside the modal for starting status', () => {
      render(
        <TransactionStatus
          {...baseProps}
          currentStatus="Starting Transaction"
        />
      );

      // Click outside the modal
      fireEvent.mouseDown(document.body);

      expect(baseProps.closeTransactionStatus).not.toHaveBeenCalled();
    });
  });

  describe('handles buy mode', () => {
    it('renders buy mode correctly', () => {
      const buyProps = {
        ...baseProps,
        isBuy: true,
        buyToken: mockBuyToken,
        sellToken: null,
        usdAmount: '50.00',
        payingTokens: mockPayingTokens,
      };

      render(
        <TransactionStatus {...buyProps} currentStatus="Starting Transaction" />
      );

      expect(screen.getByText('Starting Transaction')).toBeInTheDocument();
    });

    it('passes correct props to TransactionDetails for buy mode', () => {
      const buyProps = {
        ...baseProps,
        isBuy: true,
        buyToken: mockBuyToken,
        sellToken: null,
        usdAmount: '50.00',
        payingTokens: mockPayingTokens,
        completedTxHash: '0xcompleted123',
        completedChainId: 1,
      };

      render(
        <TransactionStatus {...buyProps} currentStatus="Transaction Pending" />
      );

      // Open details
      const viewStatusButton = screen.getByText('View Status');
      fireEvent.click(viewStatusButton);

      expect(screen.getByTestId('transaction-details')).toBeInTheDocument();
    });
  });

  describe('handles sell mode', () => {
    it('renders sell mode correctly', () => {
      render(
        <TransactionStatus
          {...baseProps}
          currentStatus="Starting Transaction"
        />
      );

      expect(screen.getByText('Starting Transaction')).toBeInTheDocument();
    });

    it('passes correct props to TransactionDetails for sell mode', () => {
      render(
        <TransactionStatus {...baseProps} currentStatus="Transaction Pending" />
      );

      // Open details
      const viewStatusButton = screen.getByText('View Status');
      fireEvent.click(viewStatusButton);

      expect(screen.getByTestId('transaction-details')).toBeInTheDocument();
    });
  });

  describe('handles different status icons and styling', () => {
    it('applies correct styling for starting transaction', () => {
      render(
        <TransactionStatus
          {...baseProps}
          currentStatus="Starting Transaction"
        />
      );

      const iconContainer = screen.getByAltText(
        'Starting Transaction'
      ).parentElement;
      expect(iconContainer).toHaveClass(
        'w-[90px]',
        'h-[90px]',
        'rounded-full',
        'border-[3px]',
        'border-white/10',
        'bg-[#8A77FF]'
      );
    });

    it('applies correct styling for transaction pending', () => {
      render(
        <TransactionStatus {...baseProps} currentStatus="Transaction Pending" />
      );

      const iconContainer = screen.getByAltText(
        'Transaction Pending'
      ).parentElement;
      expect(iconContainer).toHaveClass(
        'w-[90px]',
        'h-[90px]',
        'rounded-full',
        'border-[3px]',
        'border-white/10',
        'bg-[#8A77FF]'
      );
    });

    it('applies correct styling for transaction complete', () => {
      render(
        <TransactionStatus
          {...baseProps}
          currentStatus="Transaction Complete"
        />
      );

      const iconContainer = screen.getByAltText(
        'Transaction Complete'
      ).parentElement;
      expect(iconContainer).toHaveClass(
        'w-[90px]',
        'h-[90px]',
        'rounded-full',
        'border-[4.5px]',
        'border-[#5CFF93]',
        'bg-[#5CFF93]/30'
      );
    });

    it('applies correct styling for transaction failed', () => {
      render(
        <TransactionStatus {...baseProps} currentStatus="Transaction Failed" />
      );

      const iconContainer =
        screen.getByAltText('Transaction Failed').parentElement;
      expect(iconContainer).toHaveClass(
        'w-[90px]',
        'h-[90px]',
        'rounded-full',
        'border-[4.5px]',
        'border-[#FF366C]',
        'bg-[#FF366C]/30'
      );
    });
  });

  describe('handles edge cases', () => {
    it('handles missing optional props gracefully', () => {
      const minimalProps = {
        closeTransactionStatus: vi.fn(),
        userOpHash: '0x1234567890abcdef',
        chainId: 1,
        currentStatus: 'Starting Transaction' as const,
        errorDetails: '',
      };

      render(<TransactionStatus {...minimalProps} />);

      expect(screen.getByText('Starting Transaction')).toBeInTheDocument();
    });

    it('handles null token values', () => {
      const propsWithNullTokens = {
        ...baseProps,
        sellToken: null,
        buyToken: null,
      };

      render(<TransactionStatus {...propsWithNullTokens} />);

      expect(screen.getByText('Starting Transaction')).toBeInTheDocument();
    });

    it('handles missing sell offer', () => {
      const propsWithoutSellOffer = {
        ...baseProps,
        sellOffer: null,
      };

      render(<TransactionStatus {...propsWithoutSellOffer} />);

      expect(screen.getByText('Starting Transaction')).toBeInTheDocument();
    });
  });

  describe('handles different chain IDs', () => {
    it('renders with different chain ID', () => {
      const propsWithDifferentChain = {
        ...baseProps,
        chainId: 137, // Polygon
      };

      render(<TransactionStatus {...propsWithDifferentChain} />);

      expect(screen.getByText('Starting Transaction')).toBeInTheDocument();
    });
  });

  describe('handles error states', () => {
    it('displays error details when provided', () => {
      const propsWithError = {
        ...baseProps,
        currentStatus: 'Transaction Failed' as const,
        errorDetails: 'Transaction failed due to insufficient gas',
      };

      render(<TransactionStatus {...propsWithError} />);

      expect(screen.getByText('Transaction Failed')).toBeInTheDocument();
    });

    it('handles resource lock failure', () => {
      const propsWithResourceLockFailure = {
        ...baseProps,
        isResourceLockFailed: true,
      };

      render(<TransactionStatus {...propsWithResourceLockFailure} />);

      expect(screen.getByText('Starting Transaction')).toBeInTheDocument();
    });
  });
});
