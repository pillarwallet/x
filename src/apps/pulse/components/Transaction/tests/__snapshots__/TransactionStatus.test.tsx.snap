// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<TransactionStatus /> > renders correctly and matches snapshot 1`] = `
<div
  className="flex flex-col w-full max-w-[446px] h-[calc(100vh-94px)] max-h-[600px] bg-[#1E1D24] border border-white/5 rounded-[10px] p-6"
>
  <div
    className="flex flex-col items-center justify-center h-full"
  >
    <div
      className="w-[90px] h-[90px] rounded-full border-[3px] border-white/10 bg-[#8A77FF] flex items-center justify-center flex-shrink-0"
    >
      <img
        alt="Starting Transaction"
        className="w-[60px] h-[60px]"
        src="data:image/svg+xml,%3csvg%20width='61'%20height='60'%20viewBox='0%200%2061%2060'%20fill='none'%20xmlns='http://www.w3.org/2000/svg'%3e%3cpath%20opacity='0.4'%20d='M34.81%2048.0128C33.0167%2050.3805%2031.5555%2052.31%2030.3127%2053.5723C29.1497%2054.7538%2027.5325%2056.0378%2025.5377%2055.506C23.421%2054.942%2022.8381%2052.8993%2022.6112%2051.322C22.3699%2049.6455%2022.37%2047.312%2022.3701%2044.485V39.7178L37.0522%2044.7963L34.81%2048.0128Z'%20fill='white'/%3e%3cpath%20d='M47.1273%204.9353C49.4888%204.21403%2051.8128%204.00053%2053.7673%205.45345C55.7313%206.91348%2056.1723%209.19933%2056.1255%2011.6528C56.0798%2014.0723%2055.5413%2017.2649%2054.872%2021.2329L52.433%2035.6953C52.0003%2038.261%2051.642%2040.386%2051.124%2042.0043C50.5795%2043.705%2049.7615%2045.1735%2048.193%2046.0908C46.6293%2047.0058%2044.9438%2047.0073%2043.1833%2046.6628C41.523%2046.338%2039.4983%2045.6405%2037.0515%2044.7943L37.0345%2044.819L22.3694%2039.7158L42.9945%2017.5L16.7445%2037.7585C13.8688%2036.737%2011.5008%2035.8285%209.70523%2034.9115C7.17873%2033.621%205.02968%2031.9408%204.87748%2029.1948C4.72521%2026.4468%206.67801%2024.5426%209.04748%2022.9867C11.4594%2021.403%2015.0409%2019.7103%2019.5622%2017.5734L38.1643%208.78113C41.8403%207.04365%2044.7918%205.64855%2047.1273%204.9353Z'%20fill='white'/%3e%3c/svg%3e"
      />
    </div>
    <div
      className="text-white font-normal text-[20px] mt-3"
    >
      Starting Transaction
    </div>
    <div
      className="text-white/50 font-normal text-[16px] mt-[10px]"
    >
      Just a moment...
    </div>
  </div>
</div>
`;
